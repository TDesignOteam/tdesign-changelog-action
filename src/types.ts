import type { RestEndpointMethodTypes } from '@octokit/plugin-rest-endpoint-methods'

export type Pulls = RestEndpointMethodTypes['pulls']['get']['response']
export type PullsData = Pulls['data']
