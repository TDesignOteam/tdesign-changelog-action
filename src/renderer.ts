import { info } from '@actions/core'
import type { PRChangelog, PullsData } from './types'

const skipChangelogLabel = ['skip-changelog']
const fixLabel = ['fix', 'bug', 'hotfix']
const breakingLabel = ['break', 'breaking', 'breaking changes']
const featureLabel = ['feature', 'feat', 'enhancement']
export const CHANGELOG_REG = /-\s([A-Z]+)(?:\(([A-Z\s]*)\))?:\s(.+)/gi
export const PULL_NUMBER_REG = /in\shttps:\/\/github\.com\/.+\/pull\/(\d+)/g
export const SKIP_CHANGELOG_REG = /\[x\] 本条 PR 不需要纳入 changelog/i
export function getPullNumbers(body: string) {
  const arr = [...body.matchAll(PULL_NUMBER_REG)]
  const pullNumbers = arr.map(n => Number(n[1])) // pr number list
  const uniquePullNumbers = [...new Set(pullNumbers)]
  return uniquePullNumbers
}

function regToPrObj(arr: string[]) {
  return {
    cate: arr[1],
    component: arr[2] || '',
    desc: arr[3],
  }
}
function renderCate(cate: PRChangelog[]) {
  return `${cate.sort().map((pr) => {
          const title = pr.changelog ? `\`${pr.changelog.component || pr.changelog.cate}\`: ${pr.changelog.desc}` : pr.title
          return `- ${title} @${pr.user.login} ([#${pr.number}](${pr.html_url}))`
      }).join('\n')}`
}

export function renderMarkdown(pullRequestList: PullsData[]) {
  // 分类靠标签
  // 标题看有没有更新日志
  const categories = {
    breaking: [] as PRChangelog[],
    features: [] as PRChangelog[],
    bugfix: [] as PRChangelog[],
    extra: [] as PRChangelog[],
  }

  pullRequestList.forEach((pr) => {
    pr.body = pr.body ? pr.body : ''

    // 不需要纳入 changelog 的 label
    if (pr.labels.find(l => skipChangelogLabel.includes(l.name))) {
      info(`pr ${pr.number} 有skipChangelogLabel`)
      return
    }
    // 在 pr body 明确填了 跳过 label
    if (SKIP_CHANGELOG_REG.test(pr.body)) {
      info(`pr ${pr.number} 显示不需要纳入 changelog`)
      return
    }

    if (pr.body.includes('### 📝 更新日志')) {
      const arr = [...pr.body.matchAll(CHANGELOG_REG)]

      if (arr.length === 0) {
        info(`没有找到任何一条日志内容 number:${pr.number}, body:${pr.body}`)
        categories.extra.push(pr)
        return
      }

      arr.map(a => regToPrObj(a)).forEach((changelog) => {
        const logItem = {
          ...pr,
          changelog,
        }

        function isInLabel(label: string[]) {
          return label.includes(changelog.cate) || (arr.length === 1 && pr.labels.some(l => label.includes(l.name)))
        }

        if (isInLabel(breakingLabel)) {
          categories.breaking.push(logItem)
        }
        else if (isInLabel(featureLabel)) {
          categories.features.push(logItem)
        }
        else if (isInLabel(fixLabel)) {
          categories.bugfix.push(logItem)
        }
        else {
          categories.extra.push(logItem)
        }
      })
    }
    else {
      // 说明开发者没有按模版填写 pr，默认取 title
      info(`pr ${pr.number} 没有填写模版`)
      categories.extra.push(pr) // ??
    }
  })

  return [
    categories.breaking.length
      ? `### ❗ Breaking Changes
${renderCate(categories.breaking)}`
      : '',

    categories.features.length
      ? `### 🚀 Features
${renderCate(categories.features)}`
      : '',

    categories.bugfix.length
      ? `### 🐞 Bug Fixes
${renderCate(categories.bugfix)}`
      : '',

    categories.extra.length
      ? `### 🚧 Others
${renderCate(categories.extra)}`
      : '',
  ].filter(n => n).join('\n')
}
