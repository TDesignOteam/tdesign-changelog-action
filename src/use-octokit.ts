import { getOctokit } from '@actions/github'

export interface OctokitContext {
  token: string
  owner?: string
  repo?: string
}
export function useOctokit(context: OctokitContext) {
  const octokit = getOctokit(context.token)

  const generateReleaseNotes = async (owner: string, repo: string, tag_name: string, target_commitish: string) => {
    const res = octokit.rest.repos.generateReleaseNotes({
      owner,
      repo,
      tag_name,
      target_commitish,
    })
    return res
  }

  const getPullRequest = (owner: string, repo: string, pull_number: number) => {
    const res = octokit.rest.pulls.get({
      owner,
      repo,
      pull_number,
    })
    return res
  }

  return {
    generateReleaseNotes,
    getPullRequest,
  }
}
