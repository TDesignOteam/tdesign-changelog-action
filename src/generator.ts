import type { PullsData } from './types'
import { readFileSync } from 'node:fs'
import process from 'node:process'
import { endGroup, getInput, info, setOutput, startGroup } from '@actions/core'
import dayjs from 'dayjs'
import { getPullNumbers, renderMarkdown } from './renderer'
import { useOctokit } from './useOctokit'

export async function generatorLogStart(context) {
  if (context.payload.action === 'closed') {
    info(`[generator] PR is closed, skip`)
    return ''
  }

  const GITHUB_TOKEN = process.env.GITHUB_TOKEN || ''
  const { generateReleaseNotes, getPullRequest } = useOctokit({ token: GITHUB_TOKEN })
  let tag = getInput('tag', { required: false })
  if (!tag) {
    const pkg = JSON.parse(readFileSync('./package.json', 'utf-8'))
    tag = pkg.version
  }
  const { owner, repo } = context.repo
  info(`owner:${owner}, repo:${repo}`)

  const releases = await generateReleaseNotes(
    owner,
    repo,
    tag,
    'develop',
  )

  const PRNumbers = getPullNumbers(releases.data.body)

  const PRListRes = await Promise.all(PRNumbers.map(pull_number => getPullRequest(
    owner,
    repo,
    pull_number,
  )))

  const PRList = PRListRes.map(res => res.data as PullsData)
  startGroup(`[generator] PRList`)
  info(`PRList count:${PRList.length}`)
  info(`PRList json:${JSON.stringify(PRList)}`)
  endGroup()

  const logRelease = `(删除此行代表确认该日志): 修改并确认日志后删除这一行，机器人会提交到 本 PR 的 CHANGELOG.md 文件中
## 🌈 ${tag} \`${dayjs().format('YYYY-MM-DD')}\` \n${renderMarkdown(PRList)}\n`

  info(logRelease)

  setActionOutput(logRelease)
  return logRelease
}

function setActionOutput(changelog: string) {
  setOutput('changelog', changelog)
}
