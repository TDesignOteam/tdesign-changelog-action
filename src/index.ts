import fs from 'node:fs'
import process from 'node:process'
import { getInput, info, setFailed, setOutput } from '@actions/core'
import { context, getOctokit } from '@actions/github'
import dayjs from 'dayjs'
import { getPReformatNotes, renderMarkdown } from './renderer'
import type { PullsData } from './types'

const GITHUB_TOKEN = process.env.GITHUB_TOKEN

info(`github.context:${JSON.stringify(context)}`)

// console.log('payload', context.payload);

if (!GITHUB_TOKEN) {
  throw new Error(
    'GitHub\'s API requires a token. Please pass a valid token (GITHUB_TOKEN) as an env variable, no scopes are required.',
  )
}

const octokit = getOctokit(GITHUB_TOKEN)

async function generatorLogStart() {
  let tag = getInput('tag', { required: false })
  if (!tag) {
    const pkg = JSON.parse(fs.readFileSync('./package.json', 'utf-8'))
    tag = pkg.version
  }
  const { owner, repo } = context.repo
  info(`owner:${owner}, repo:${repo}`)

  const releases = await octokit.rest.repos.generateReleaseNotes({
    owner,
    repo,
    tag_name: tag, // 'package.version'
    target_commitish: 'develop', // 也可以从上下文中拿
  })

  const PRNumbers = getPReformatNotes(releases.data.body)

  const PRListRes = await Promise.all(PRNumbers.map(pull_number => octokit.rest.pulls.get({
    owner,
    repo,
    pull_number,
  })))

  const PRList = PRListRes.map(res => res.data as PullsData)

  info(`PRList:${JSON.stringify(PRList)}`)

  const logRelease = `(删除此行代表确认该日志): 修改并确认日志后删除这一行，机器人会提交到 本 PR 的 CHANGELOG.md 文件中
## 🌈 ${tag} \`${dayjs().format('YYYY-MM-DD')}\` \n${renderMarkdown(PRList)}\n`

  info(logRelease)

  setActionOutput(logRelease)
  return logRelease
}

generatorLogStart().catch((error) => {
  console.error(error)
  setFailed(`💥 Auto Release failed with error: ${error.message}`)
})

function setActionOutput(changelog: string) {
  setOutput('changelog', changelog)
}
