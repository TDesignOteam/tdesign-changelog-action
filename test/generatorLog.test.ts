import { describe, expect, it, vi } from 'vitest'
import { context } from '@actions/github'
import { generatorLogStart } from '../src/generator'

describe('generatorLog', () => {
  it(': generatorLogStart', async () => {
    const mockDate = new Date(2024, 1, 24)
    vi.setSystemTime(mockDate)
    const log = await generatorLogStart(context)
    expect(log).toMatchSnapshot()
    vi.useRealTimers()
  })
})
