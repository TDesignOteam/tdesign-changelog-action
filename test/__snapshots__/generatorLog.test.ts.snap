// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generatorLog > : generatorLogStart 1`] = `
"(删除此行代表确认该日志): 修改并确认日志后删除这一行，机器人会提交到 本 PR 的 CHANGELOG.md 文件中
  ## 🌈 0.2.0 \`2024-02-24\` 
### ❗ Breaking Changes
- \`C\`: a major update label in break @liweijie0812 ([#29](https://github.com/TDesignOteam/tdesign-changelog-action/pull/29))
- \`C\`: a major update label in breaking @liweijie0812 ([#29](https://github.com/TDesignOteam/tdesign-changelog-action/pull/29))
- \`g\`: a minor fix label in break @liweijie0812 ([#29](https://github.com/TDesignOteam/tdesign-changelog-action/pull/29))
- \`H H\`: a major update label in break @liweijie0812 ([#29](https://github.com/TDesignOteam/tdesign-changelog-action/pull/29))
### 🚀 Features
- \`B\`: a feature label in feat @liweijie0812 ([#29](https://github.com/TDesignOteam/tdesign-changelog-action/pull/29))
- \`B\`: a feature label in feature @liweijie0812 ([#29](https://github.com/TDesignOteam/tdesign-changelog-action/pull/29))
- \`B\`: a feature label in enhancement @liweijie0812 ([#29](https://github.com/TDesignOteam/tdesign-changelog-action/pull/29))
- \`D\`: a feature  label in feat @liweijie0812 ([#29](https://github.com/TDesignOteam/tdesign-changelog-action/pull/29))
### 🐞 Bug Fixes
- \`A\`: a bug fix label in fix @liweijie0812 ([#29](https://github.com/TDesignOteam/tdesign-changelog-action/pull/29))
- \`A\`: a bug fix label in hotfix @liweijie0812 ([#29](https://github.com/TDesignOteam/tdesign-changelog-action/pull/29))
- \`A\`: a bug fix label in bug @liweijie0812 ([#29](https://github.com/TDesignOteam/tdesign-changelog-action/pull/29))
### 🚧 Others
- \`docs\`: a doc change label in docs @liweijie0812 ([#29](https://github.com/TDesignOteam/tdesign-changelog-action/pull/29))
- \`chore\`: a chore label in chore @liweijie0812 ([#29](https://github.com/TDesignOteam/tdesign-changelog-action/pull/29))
- \`I\`: a refactor label in refactor @liweijie0812 ([#29](https://github.com/TDesignOteam/tdesign-changelog-action/pull/29))
"
`;
